Fix: Fails to import packages where declared license is multiline

tuxdna opened the pull request:
Fix for: https://github.com/fabric8-analytics/fabric8-analytics-data-model/issues/71

tuxdna commented:
@msrb @jpopelka PTAL :-)

tuxdna commented:
@sunilk747 You might also want to take a look as this PR and https://github.com/fabric8-analytics/fabric8-analytics-data-model/pull/70, both seems related

miteshvp commented on src/graph_populator.py:
```diff
@@ -106,7 +106,21 @@ def construct_version_query(cls, input_json):
+                    'Common Development And Distribution License (CDDL) Version 1.0']
+                    """
+                    s = details[0]['declared_license']
+                    declared_licenses = list(map(lambda x: x.strip(), s.split("\n")))
```
@tuxdna - WDYT about rather than splitting based on newline, you join them and split by `,` to be consistent

tuxdna replied:
@miteshvp That doesn't make any difference to the output.

sunilk747 replied:
@miteshvp @tuxdna IMO Let,s join them rather than splitting it on newline.After that we can use `,` and `and` as delimiter for splitting. In this way we can have more consistent list of licenses which will be good fit for our license analysis platform.

tuxdna replied:
This PR is for the case where we know that declared_license is a `str` and not a collection. So it doesn't make any sense to join it. Besides the target use-case doesn't have a comma.

Please check comments before the changeset.

tuxdna replied:
@sunilk747 If you think it is better to cover this use-case in your other PR https://github.com/fabric8-analytics/fabric8-analytics-data-model/pull/70 , please feel free to do so.

miteshvp replied:
@tuxdna - what I meant by join was replace `\n` with a `space` and then do the split operation

sunilk747 replied:
IMO in cases where the declared license is a string we should have a consistent delimiter to separate out licenses, that way it would be easy to perform license analysis.

jpopelka replied:
I'd suggest using [list comprehensions](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions) instead of `list(map(lambda ...`, i.e. in this case
```python
declared_licenses = [x.strip() for x in s.split("\n")]
```

tuxdna replied:
@jpopelka Agreed, comprehensions are more readable.

tuxdna commented:
@sunilk747 @miteshvp 

Your description doesn't quite fit the failure case. Check the output below

```python
In [32]: s
Out[32]: 'Apache License, Version 2.0 and\n        Common Development And Distribution License (CDDL) Version 1.0'

In [33]: no_newlines = " ".join(map(lambda x: x.strip(), s.split("\n")))

In [34]: no_newlines
Out[34]: 'Apache License, Version 2.0 and Common Development And Distribution License (CDDL) Version 1.0'

In [35]: declared_licenses = list(map(lambda x: x.strip(), no_newlines.split(",")))

In [36]: declared_licenses
Out[36]: 
['Apache License',
 'Version 2.0 and Common Development And Distribution License (CDDL) Version 1.0']

```

The resulting `declared_licenses` are not what they should be. They are in-fact separated by `and\n` instead of `,`

sunilk747 commented:
@tuxdna  We need something similar to this. WDYT?
```
In [9]: import re

In [10]: s = 'Apache License, Version 2.0 and\n        Common Development And Distribution License (CDDL) Version 1.0'

In [11]: no_newlines = " ".join(map(lambda x: x.strip(), s.split("\n")))

In [12]: no_newlines
Out[12]: 'Apache License, Version 2.0 and Common Development And Distribution License (CDDL) Version 1.0'

In [13]: licenses = re.split('; |, | and', no_newlines)

In [14]: licenses
Out[14]: 
['Apache License',
 'Version 2.0',
 ' Common Development And Distribution License (CDDL) Version 1.0']

In [15]: final_licenses = list()

In [16]: for l in licenses:
    ...:     if l.startswith('Version'):
    ...:         final_licenses[-1] = final_licenses[-1] + l
    ...:     else:
    ...:         final_licenses.append(l)
    ...:         

In [17]: final_licenses
Out[17]: 
['Apache LicenseVersion 2.0',
 ' Common Development And Distribution License (CDDL) Version 1.0']
```

tuxdna commented:
@sunilk747 @jpopelka @miteshvp Updated the code. Please check.

sunilk747 commented on src/graph_populator.py:
```diff
@@ -111,16 +111,29 @@ def construct_version_query(cls, input_json):
 
                     ['Apache License, Version 2.0 and',
                     'Common Development And Distribution License (CDDL) Version 1.0']
+
```
@tuxdna If the string is `"Apache License, Version2, GPL, Version 2.1"` from your recent changes the output would be `['Apache License', 'Version2', 'GPl', 'Version 2.1']` which is not correct. If you could do something similar to this `['Apache License, Version2', 'GPl, Version 2.1']` it would be great.

tuxdna replied:
@sunilk747 In the first place this shouldn't be the case because `declared_license` should only contain a single license entry. Fix in this PR is to avoid failures that are happening due to `\n`.

Now in free text like this, the licences could written in any way. So is becomes a very context specific problem and it will require more sophisticated approaches, which I believe is beyond the scope of this problem. You may want to create a new issue for this use-case.

msrb commented:
What's weird is that the package that caused the failure actually doesn't contain the new line in pom.xml: http://central.maven.org/maven2/org/apache/tomcat/tomcat-servlet-api/8.5.14/tomcat-servlet-api-8.5.14.pom 

The new line only appears in our data, from some reason.

tuxdna commented:
@msrb If you view source at this location http://central.maven.org/maven2/org/apache/tomcat/tomcat-servlet-api/8.5.14/tomcat-servlet-api-8.5.14.pom you will notice the newline.

msrb commented:
Ah, I see it now, firefox changes the formatting. `curl` doesn't lie.

msrb commented:
LGTM, thanks @tuxdna 😉
